### ISSUE TITLE
Using Emoji causes twice the memory usage on iOS

### ISSUE THREAD

**Issue by ldhlfzysys** [flutter/flutter#69566]

**Comment by ldhlfzysys** [comment]
Scaffold(
      backgroundColor: Colors.white,
      appBar: AppBar(
        elevation: 0.0,
        title: Text('test'),
      ),
      body: Container(
        child: Text('test'),
      ));

![image](https://user-images.githubusercontent.com/7666564/97862522-7734e600-1d40-11eb-900a-be3fe4772790.png)


Scaffold(
      backgroundColor: Colors.white,
      appBar: AppBar(
        elevation: 0.0,
        title: Text('点评讨论'),
      ),
      body: Container(
        child: Text('🎬'),
      ));

![image](https://user-images.githubusercontent.com/7666564/97862600-93388780-1d40-11eb-95c8-8ffdf468bb3c.png)

**Comment by TahaTesser** [comment]
Hi @ldhlfzysys 
Can you please provide your `flutter doctor -v` your `flutter run --verbose` and a complete reproducible minimal code sample
Thank you

**Comment by ldhlfzysys** [comment]
[✓] Flutter (Channel stable, 1.22.3, on Mac OS X 10.15.5 19F96, locale zh-Hans-CN)

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.1)
[✓] Xcode - develop for iOS and macOS (Xcode 12.0.1)
[✓] Android Studio (version 4.0)
[✓] VS Code (version 1.50.1)
[✓] Connected device (2 available)

• No issues found!

**Comment by ldhlfzysys** [comment]
Text('🎬')

**Comment by AlexV525** [comment]
Can you provide more info about what kind of devices you have tested?

**Comment by ldhlfzysys** [comment]
macos, iPhone 11 device and iPhone 11 simulator

**Comment by ldhlfzysys** [comment]
All emoji have this problem

**Comment by TahaTesser** [comment]
Hi @ldhlfzysys 
I can reproduce the memory issue using other emojis as well 

<details> 
<summary>code sample</summary> 

```dart
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Material App',
      home: Home(),
    );
  }
}

class Home extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
        backgroundColor: Colors.white,
        appBar: AppBar(
          elevation: 0.0,
          title: Text('点评讨论'),
        ),
        body: Container(
          child: Text('🔥'),
        ));
  }
}

```

</details>

<details> 
<summary>flutter doctor -v</summary> 

```bash
[✓] Flutter (Channel stable, 1.22.3, on Mac OS X 10.15.7 19H2, locale en-GB)
    • Flutter version 1.22.3 at /Users/tahatesser/Code/flutter_stable
    • Framework revision 8874f21e79 (5 days ago), 2020-10-29 14:14:35 -0700
    • Engine revision a1440ca392
    • Dart version 2.10.3

 
[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
    • Android SDK at /Users/tahatesser/Code/sdk
    • Platform android-30, build-tools 30.0.2
    • ANDROID_HOME = /Users/tahatesser/Code/sdk
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 12.1)
    • Xcode at /Volumes/Extreme/Xcode.app/Contents/Developer
    • Xcode 12.1, Build version 12A7403
    • CocoaPods version 1.10.0

[!] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)

[✓] VS Code (version 1.50.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.16.0

[!] Connected device
    ! No devices available

! Doctor found issues in 2 categories.
```
```bash
[✓] Flutter (Channel master, 1.24.0-7.0.pre.73, on Mac OS X 10.15.7 19H2
    darwin-x64, locale en-GB)
    • Flutter version 1.24.0-7.0.pre.73 at /Users/tahatesser/Code/flutter_master
    • Framework revision fb90544cab (2 hours ago), 2020-11-02 22:48:03 -0800
    • Engine revision 3460519398
    • Dart version 2.12.0 (build 2.12.0-3.0.dev)

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
    • Android SDK at /Users/tahatesser/Code/sdk
    • Platform android-30, build-tools 30.0.2
    • ANDROID_HOME = /Users/tahatesser/Code/sdk
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 12.1)
    • Xcode at /Volumes/Extreme/Xcode.app/Contents/Developer
    • Xcode 12.1, Build version 12A7403
    • CocoaPods version 1.10.0

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)

[✓] VS Code (version 1.50.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.16.0

[✓] Connected device (4 available)
    • Taha’s iPad (mobile) • 00008020-000255113EE8402E • ios            • iOS
      14.1
    • macOS (desktop)      • macos                     • darwin-x64     • Mac OS
      X 10.15.7 19H2 darwin-x64
    • Web Server (web)     • web-server                • web-javascript •
      Flutter Tools
    • Chrome (web)         • chrome                    • web-javascript • Google
      Chrome 86.0.4240.183
    ! Error: Taha’s iPad is busy: Copying cache files from device. Xcode will
      continue when Taha’s iPad is finished. (code -10)

• No issues found!
```

</details>

Related https://github.com/flutter/flutter/issues/42586

**Comment by ldhlfzysys** [comment]
@TahaTesser  Hi~ Is this problem improved in version 1.25?

**Comment by pouloghost** [comment]
How's fixing going, we have several new features blocked by this issue. REALLY NEEDS FIX!

**Comment by xiushaomin** [comment]
urgently need to solve this problem，The Flutter engine takes up enough memory~

**Comment by xiushaomin** [comment]
I found the reason for this problem in font_skia.cc,  in the `gteTableSize(tag)` method, the Apple Color Emoji's sbix table is too large，This sbix table contains all emoji png data, so the sbix table_szie have 120mb, 

```   
const size_t table_size = typeface->getTableSize(tag);
  if (table_size == 0)
    return nullptr;
  void* buffer = malloc(table_size);
  if (buffer == nullptr)
    return nullptr;

````
i also found the sbix table，just use to layout？only the methods `is_color_bitmap_font` and `hb_font_get_glyph_extents(hbFont, glyph_ix, &extents)` use the sbix table blob. is there any way to avoid malloc sbix table? @TahaTesser @AlexV525 @chinmaygarde

**Comment by zanderso** [comment]
@jason-simmons for thoughts on this, and whether this issue will be affected by https://github.com/flutter/flutter/issues/39420.

**Comment by jason-simmons** [comment]
I looked into emoji font memory usage in Libtxt and SkParagraph.

### Libtxt

Libtxt provides font tables to HarfBuzz through the `GetTable` callback in `font_skia.cc`.  This function allocates a buffer for the table and calls `SkTypeface::getTableData` to copy the data.

HarfBuzz APIs (such as the ones mentioned in https://github.com/flutter/flutter/issues/69566#issuecomment-755936926) may call the callback to request the SBIX table.  On my iPhone this table had a size of 57MB.

Most of that data is bitmaps for the emoji glyphs.  However, the buffer allocated for use by HarfBuzz is just used to obtain font properties during layout and is not used during painting.  Skia's Mac/iOS font implementation will use Apple font APIs to get the emoji bitmaps used for painting.

If I patch `GetTable` to return null for the SBIX table, then the emoji glyphs basically seem to render as expected in a simple test app.
I'm not sure what might go wrong if Libtxt did not provide this table to HarfBuzz/Minikin.  If there is data in the table that is required to do some layouts correctly, then Libtxt would need some other way to provide that data to HarfBuzz.

### SkParagraph

By default SkShaper's `create_hb_face` will call `SkTypeface::openStream` and provide a copy of the font asset to HarfBuzz (including the SBIX table).  So SkParagraph's memory usage for the emoji font will be similar to Libtxt.

`create_hb_face` does have another code path where it will provide a callback to HarfBuzz that will copy tables out of the `SkTypeface` on demand (similar to how Libtxt works).  I tried forcing usage of this mode and found that the SBIX table was not loaded by HarfBuzz.  So this saved a significant amount of memory.

@bungeman What would happen if SkShaper always used `hb_face_create_for_tables`/`skhb_get_table` when building HarfBuzz faces?

Or would it make sense to have SkShaper do this selectively based on the size of potentially unneeded tables like SBIX in the SkTypeface?

**Comment by bungeman** [comment]
I'm confused. Using `hb_face_create_for_tables` will force a copy for each table loaded (and each time `SkTypeface::copyTableData` is called). `hb_face_create` itself should not create copies. When using an `SkTypeface` backed by a system `CTFont` the first time `SkTypeface::openStream` is called it will need to load that font into accessible memory and will keep hold of that data for future use. Is this the copy of `sbix` you're seeing? If so, this is something of a design tradeoff due to CoreText not providing an API to get at all of the font data (before `SkTypeface::makeClone` was added holding on to this data was really important, I'd have to check to see if that is no longer the case). If instead you're seeing HarfBuzz making copies of the data I'd be interested to know where.

Using `hb_face_create_for_tables` would be quite bad for the other systems like FreeType, DirectWrite, and even CTFonts created from data. These currently should all be zero copy.

There is information in the `sbix` table which should be used for correct layout. If you don't provide an `sbix` table HarfBuzz will just use whatever is in the `glyf` table which may or may not work in expected ways, depending on how the font is structured.

I'm not sure if it would help, but in theory `Paragraph` could be modified to use `SkShaper::MakeCoreText`. I'm not sure if that has been tried, but the CoreText shaper has system level access directly into the internal system font table cache and can avoid making any copies at all. Of course it is the system shaper, so will give slightly different results.

**Comment by jason-simmons** [comment]
Yes - calling `SkTypeface::openStream` loads the entire font asset into memory (around 58MB for the Apple emoji font in my test setup).  Most of the 58MB is the contents of the `sbix` table.

The `hb_face_create_for_tables` approach avoids making a copy of `sbix` because HarfBuzz did not request that particular table in the SkShaper test scenarios that I tried.  I don't know if there are other cases in which HarfBuzz would need `sbix`.

**Comment by bungeman** [comment]
One can think of `SkTypeface::openStream` as being greedy but shared and `SkTypeface::copyTableData` as being lazy but unshared. For all fonts other than CoreText system fonts we already have access to the bytes so in those cases greedy works fine and also provides sharing. In theory `SkTypeface::copyTableData` could be shared (the typeface could hold on to the table data) or `SkTypeface::openStream` could be made less greedy.

Technically it would be possible to do something like have a stream backed by a `userfaultfd` like mechanism. This would give the best user experience since it would be both a flat memory backed view but only commit memory as needed. I don't know that macOS has a mechanism like that.

Short of that we might be able to internally signal if `SkTypeface::openStream` is inexpensive or not. We could swap over to `hb_face_create_for_tables` if it looks like it will be expensive (there isn't already a stream available and the created stream would need more than a megabyte for backing, or something like that).

Longer term we're working on a new OpenType color font format (COLRv1) which is vector and gradient based instead of bitmap based. While it isn't a short term solution, eventually we hope this font will no longer need to store gradient filled images at 2048x2048 in png format.

**Comment by yisibl** [comment]
> Longer term we're working on a new OpenType color font format (COLRv1) which is vector and gradient based instead of bitmap based. While it isn't a short term solution, eventually we hope this font will no longer need to store gradient filled images at 2048x2048 in png format.

@bungeman CLORv1 is very attractive, FreeType already supports it, can flutter use the latest FreeType to render?
In addition, does flutter now support CLORv0? I submitted a separate issue https://github.com/flutter/flutter/issues/79365

**Comment by cngaofei** [comment]
How's it going? I hope it can be solved ASAP~

**Comment by Realank** [comment]
Hope to be resolved too

**Comment by iBcker** [comment]
Any update on this?

**Comment by knopp** [comment]
Also I'm not quite sure how `CLORv0` is meant to resolve the issue with Apple Color Emoji, which is entirely bitmap based?

**Comment by cj3479** [comment]
I fix it 
this is my code

font_skia.cc
![image](https://user-images.githubusercontent.com/9412601/148183625-12ae3bde-f876-4216-8c30-6f2b0e11b6c3.png)

**Comment by tinyc-z** [comment]
> I fix it this is my code
> 
> font_skia.cc ![image](https://user-images.githubusercontent.com/9412601/148183625-12ae3bde-f876-4216-8c30-6f2b0e11b6c3.png)

could you create a pull request, i think it will help more people?

**Comment by eggfly** [comment]
> I fix it this is my code
> 
> font_skia.cc ![image](https://user-images.githubusercontent.com/9412601/148183625-12ae3bde-f876-4216-8c30-6f2b0e11b6c3.png)

could you create a pull request, i think it will help more people? @tinyc-z 💪

**Comment by zanderso** [comment]
@jason-simmons @bungeman @knopp any thoughts about the patch in https://github.com/flutter/flutter/issues/69566#issuecomment-1005470784?

**Comment by knopp** [comment]
IIRC I tried similar approach where I just disregarded any `SBIX` table and it seemed to work. It didn't look like any of the font layout got broken. But I didn't test it thoroughly.

**Comment by jason-simmons** [comment]
https://github.com/flutter/flutter/issues/69566#issuecomment-769575070 discusses the potential effects of ignoring the SBIX table.

The https://github.com/flutter/flutter/issues/69566#issuecomment-1005470784 patch affects font table loading in Libtxt, and we are planning to migrate to SkParagraph soon which uses a different code path.

For SkShaper/SkParagraph I'd be interested in exploring ways to avoid the `SkTypeface::openStream` call for expensive fonts and instead load tables on demand (see https://github.com/flutter/flutter/issues/69566#issuecomment-770036791)

**Comment by knopp** [comment]
I had a fix for this in mind where instead of ignoring the SBIX table we would process it, preserving layout information but stripping bitmap data (which is not used during rasterization). But this is not really my domain so perhaps there are better ways of solving this.

**Comment by Rusino** [comment]
I think this one is fixed. Could we confirm that?

**Comment by jason-simmons** [comment]
This has been fixed in SkParagraph.
